lib/graphql/stitching/shaper.rb

```diff
@@ -54,21 +58,35 @@ def munge_entry(entry, selections, parent_type)
 
       def munge_field(entry, node, parent_type)
         field_identifier = (node.alias || node.name)
-        node_type = Util.get_named_type_for_field_node(@supergraph.schema, parent_type, node)
+        named_type = Util.get_named_type_for_field_node(@supergraph.schema, parent_type, node)
+        field_type = parent_type.own_fields[node.name].type
 
         if entry.nil?
-          return  # TODO bubble up error if not nullable
+          raise InvalidNullError.new(parent_type, parent_type.own_fields[node.name], nil) if field_type.non_null?
+          return
         end
 
         child_entry = entry[field_identifier]
         if child_entry.nil?
+          raise InvalidNullError.new(parent_type, parent_type.own_fields[node.name], nil) if field_type.non_null?
           entry[field_identifier] = nil
         elsif child_entry.is_a? Array
```

Comment:
Arrays are tricky, and there are a few issues here –

1. we can't assess array-ness based on value, because the value isn't guaranteed. It must be recognized via schema.
2. Arrays can also be hidden inside of non-null type wrappers, so you need to do some drilling to recognize array-ness in schema. Just asking `node.list?` misses a list wrapped in a non-null wrapper.
3. Then, lists can be nested, so require their own recursion to unpack and assess nullability extending all the way down to the element.